File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
    
        int mini=INT_MAX,maxi=INT_MIN;
        int profit=0;
        for(auto val: prices){
            if(maxi>val+fee){
                profit+=max(0,maxi-mini-fee);
                mini=maxi=val;               

            }else{
                mini=min(mini,val);
                maxi=max(maxi,val);
            }
            
        }
        profit+=max(0,maxi-mini-fee);
        return profit;
        
    }
};

File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution(object):
    def maxProfit(self, prices, fee):
        n=len(prices)
        minv,maxv=1000000,0
        maxprofit=0
        for i in range(n):
            if maxv > prices[i]+fee:
                maxprofit+=max(0,maxv-minv-fee)
                minv=maxv=prices[i]
            else:
                minv=min(minv,prices[i])
                maxv=max(maxv,prices[i])
        maxprofit+=max(0,maxv-minv-fee)
        return maxprofit